### DISCUSSION TITLE
Fully replacing the structure/values of a pattern

### DISCUSSION THREAD

**Discussion by larkob** [tidalcycles/strudel#155] [Q&A] [answered]
I've learned that there's plenty of options to combine two patterns (squeeze, setOut, etc.), and after trying out every one of them, I still cannot find a way to fully replace the "left" pattern with the "right" pattern while keeping their context (i.e. gain, velocity, sound, etc.) intact. Let's say I start with `s("bd")`, so I have a very simple pattern. Then I apply a rhythmic structure like this `s("bd").struct("~ x")`. Now I want to fully replace the temporal structure with another structure, but I find that to be impossible with anything I append: `s("bd").struct("~ x").structAll("x x").webdirt()`.

For patterns with values, I run into the same problem: `"~ a2" .setOut("g2 f2") .note() .piano() .out()`

It seems like once a pattern has a null value (rest/silence), it cancels the event in the hap (is that the right terminology?) and thus cannot be modified with subsequent functions. Am I missing something here, or is this a missing piece of functionality?

As a side note, I noticed that the `append` function is still mentioned in the tutorial although it no longer exists.

**Comment by felixroos** [comment]
Yep, it seems to be a problem of how silence works. Let me trace the code to understand what happens here:

1. [`silence`](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L1214) will always return an empty array for any query
2. [`struct` is a shorthand for keepifOut](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L1165), 
  2.1 ...which will (like any operator) [turn into](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L1141) `pat._opOut(other, /*..*/)`. 
  2.2 [`_opOut`](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L414) uses [`fmap`](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L198), which is an alias for `Pattern#withValue` 
  2.3 [`Pattern#withValue`](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/pattern.mjs#L192) queries the pattern and maps over the resulting haps, calling `Hap.withValue`
  2.4 [`Hap.withValue`](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/hap.mjs#L48) just [applies the given function to its value](https://github.com/tidalcycles/strudel/blob/a492628b3329337c2dcdf7d1586dcd0f7cece01a/packages/core/hap.mjs#L48)

So because silence will return an empty array for any query (1), fmap won't get a chance to map over silenced time slices (2.3).
I also created an [example](https://strudel.tidalcycles.org?RDyvc3SOo6kX) for this.

I am curious about your (musical) use case because I haven't yet had the need to restructure a pattern like this.

@yaxu I wonder if that limitation also applies to tidal. One possible way to fix this would be to change silence to return a special hap of type silence, which will be ignored by the output. With that, each slice of time would always be covered by a hap.

**Comment by felixroos** [comment]
> As a side note, I noticed that the append function is still mentioned in the tutorial although it no longer exists.

good catch, the documentation still needs some TLC! If you want to help out, feel free to send a PR. There are plenty of other things missing, see [this list](https://github.com/tidalcycles/strudel/issues/58).
I am currently into implementing soundfonts but I will come back to the doc after that!

**Comment by yaxu** [comment, marked as answer]
In Haskell there is `const` which just returns it's first argument (in js that would be `(a,b) => a`), allowing e.g. `every 3 (const "d e f) "a ~"`.

Similarly we could add a function to Pattern in strudel that just returns its argument `Pattern.prototype.replace = a => a`, then do `"~ a2" .replace("g2 f2") .note() .piano() .out()`

I think that would work for your example case but I have a feeling it's not really what you're getting at, when you mention the problem of losing context. For example:
```js
n("a ~").gain(1.1).every(3, setOut(n("b b")))
```
That would lose the final `b`.. Because there are no events in the second half of the cycle to replace with the events in the 'b' pattern.

Yep @felixroos  this is a problem in Tidal too: https://github.com/tidalcycles/Tidal/issues/721

With a related issue around having 'structured silence', that is closed not because it's really solved but because boolean patterns offer a workaround: https://github.com/tidalcycles/Tidal/issues/250

The problem is that a lot of the time, the current behaviour is what you want.. To the point that it's easy to forget about this issue. But still it, would be nice to have the option, if it doesn't over-complicate things.

**Reply by larkob** [reply]
I think I understand why this isn't working as expected. The example with the `showWhole` was very helpful for me to inspect what's going on. So far, I've used `drawLine` but I typically end up with some non-descript "object" in the console, so that's not helpful. I see now that silence/rest isn't a thing, it's the absence of anything, so it cannot be transformed or replaced. 

Perhaps we can find a different solution for what I'm trying to do: I want to play a regular pattern and then every 4 cycles I want to play a different pattern. So far, I've first defined the pattern and then applied the `every` function to say what should be played instead. What if I could use the `every` function in combination with an `everyOther` or `allBut` function that applies its function every time BUT the nth time, like this `s("bd")  .every(4,struct("~ x ~ x")) .except(4,struct("x x x x"))`? Or perhaps it's easier to understand with an `everyBut` function and the other way around: `s("bd")  .everyBut(4,struct("x x x x")) .every(4,struct("~ x ~ x"))`.

**Reply by felixroos** [reply]
[This is how i'd do that](https://strudel.tidalcycles.org/#cygiPFtiZCBzZF0hMyBbfiBiZCB%2BIGJkXT4iKS53ZWJkaXJ0KCk%3D). Having every not trigger the first time is also something i'd like, here is [one possible workaround](https://strudel.tidalcycles.org?qk6JW1Bmi26s)

**Reply by yaxu** [reply]
To finish what I was saying about `const`, here's how to do that in strudel:
```js
s("bd sd").every(4, x => s("cp*3")).webdirt()
```
If you want to be specific about which cycles you want, `when` is good for that:
```js
s("bd sd").when("<0!3 1>", x => s("cp*3")).webdirt()
```

**Reply by larkob** [reply]
Alright, a combination of `every` and `when` is pretty much what I was looking for:

```s("bd").webdirt().when("<0 1!3>", struct(pattern1).every(4, struct(pattern1))```

It's a bit counter-intuitive that the cycles begin playing the `every` pattern before the `when` pattern, but at least this approach would allow me to play any patterns, including `pattern1="~"`.
